draft-normington-jsonpath-latest.xml

```diff
@@ -218,6 +225,94 @@ dot-child-name = TBD
                   </t>
                </section>       
             </section>       
+            <section numbered="true" toc="default">
+               <name>Union Matcher</name>
+               <section numbered="true" toc="default">
+                  <name>Syntax</name>
+                  <t>A union matcher consists of one or more union elements.
+                  </t>
+                  <sourcecode type="abnf">
+matcher =/ union
+union = %x5B ws union-elements ws %x5D ; [...]
+ws = *%x20                             ; zero or more spaces
+union-elements = union-element /
+                 union-element ws %x2C ws union-elements
+                                       ; ,-separated list
+                  </sourcecode>
+               </section>       
+               <section numbered="true" toc="default">
+                  <name>Semantics</name>
+                  <t>A union matches any node which is matched by at least one of the union elements and selects the concatenation of the
+                     lists (in the order of the matchers) of descendants selected by the union elements.
+                  </t>
+               </section>
+               <section numbered="true" toc="default" anchor="unionchild">
+                  <name>Union Child</name>
+                  <section numbered="false" toc="exclude">
+                     <name>Syntax</name>
+                     <t>A union child is a union element consisting of a quoted string.
+                     </t>
+                     <sourcecode type="abnf">
+union-element = union-child ; see below for more alternatives
+union-child = %x27 *single-quoted %x27 / ; 'string'
+              %x22 *double-quoted %x22   ; "string"
```

Comment:
Maybe this one should refer to JSON string definition as of [RFC 8259.7](https://tools.ietf.org/html/rfc8259#section-7)? Otherwise we can get duplicated non-terminals if we allow JSON structures in filters.

Reply:
Actually, I'm inclined to switch to RFC 8259 syntax for strings for consistency alone. But I think we need to allow for single-quoted as well as double-quoted strings, for example like this:
```
union-child = %x22 *double-quoted %x22 / ; "string"
              %x27 *single-quoted %x27   ; 'string'

double-quoted = dq-unescaped /
          escape (
              %x22 /          ; "    quotation mark  U+0022
              %x5C /          ; \    reverse solidus U+005C
              %x2F /          ; /    solidus         U+002F
              %x62 /          ; b    backspace       U+0008
              %x66 /          ; f    form feed       U+000C
              %x6E /          ; n    line feed       U+000A
              %x72 /          ; r    carriage return U+000D
              %x74 /          ; t    tab             U+0009
              %x75 4HEXDIG )  ; uXXXX                U+XXXX


      dq-unescaped = %x20-21 / %x23-5B / %x5D-10FFFF

single-quoted = sq-unescaped /
          escape (
              %x27 /          ; '    apostrophe      U+0027
              %x5C /          ; \    reverse solidus U+005C
              %x2F /          ; /    solidus         U+002F
              %x62 /          ; b    backspace       U+0008
              %x66 /          ; f    form feed       U+000C
              %x6E /          ; n    line feed       U+000A
              %x72 /          ; r    carriage return U+000D
              %x74 /          ; t    tab             U+0009
              %x75 4HEXDIG )  ; uXXXX                U+XXXX

      sq-unescaped = %x20-26 / %x28-5B / %x5D-10FFFF

escape = %x5C                 ; \
```
@cburgmer this is the first time we would be deviating from the syntax in Proposal A. WDYT?

Reply:
> I think we need to allow for single-quoted as well as double-quoted strings

Of course, but that can be done like this:
```
union-child = %x27 *single-quoted %x27 /  ; 'string'
              json-string ; see RFC blah-blah-blah
```
I mean, JSON string already includes double-quotes in its definition.

Reply:
I think it would be better to keep the ABNF self-contained:
- It's easier to read without jumping over to another RFC. (We could make the production names the same if that helps and have a comment "copied from RFC nnnn")
- I'd like our text to be immutable, but the ABNF in another RFC could mutate because of errata (which we wouldn't _necessarily_ want to adopt, at least not without our own errata)
- it should be possible to extract the ABNF and generate a lexer/parser from it.

Reply:
@cabo Have you any general advice about whether it is best to inline or refer to ABNF from existing RFCs?

Reply:
Funny, I just answered the same question on a different list :-)

Generally, right now, it is better to copy, because there is no actionable way to refer.
Until that get's unwieldy, so if you need the date-time production from RFC 3339, that carries so much baggage you'd better refer.

We are solving the same problem for CDDL right now in the CBOR WG, so maybe we'll have something that will spill over to ABNF as wel.  For now, copy over refer, I'd say.

Reply:
I'm going to stick with the copy of the ABNF and also preserve the symmetry between single and double quoted strings as I think that's a lot more readable.